src/coreclr/vm/jithelpers.cpp

```diff
@@ -5358,17 +5366,48 @@ NOINLINE static void JIT_ReversePInvokeEnterRare(ReversePInvokeFrame* frame)
 
     frame->currentThread = thread;
 
+#ifdef PROFILING_SUPPORTED
+        if (CORProfilerTrackTransitions())
+        {
+            ProfilerUnmanagedToManagedTransitionMD(frame->pMD, COR_PRF_TRANSITION_CALL);
+        }
+#endif
+
     thread->DisablePreemptiveGC();
 }
 
 NOINLINE static void JIT_ReversePInvokeEnterRare2(ReversePInvokeFrame* frame)
 {
     frame->currentThread->RareDisablePreemptiveGC();
-}
+#ifdef DEBUGGING_SUPPORTED
+    // If the debugger is attached, we use this opportunity to see if
+    // we're disabling preemptive GC on the way into the runtime from
+    // unmanaged code. We end up here because
+    // Increment/DecrementTraceCallCount() will bump
+    // g_TrapReturningThreads for us.
+    if (CORDebuggerTraceCall())
```

Comment:
Do we have a test for this that is being fixed by this change?

I think it is too late for the TraceCall by now. We have executed the entrypoint already. If the debugger needs a hint to start stepping, I think that the hint needs to point to return address of the callback.

Reply:
We do not have a test that is fixed by this change. I pulled this from the old emitted assembly stub because I wanted to start with everything 1:1 from the old stub done in the new stub. I’m willing to move this somewhere else/remove it based on what you think is best.

Reply:
The diagnostic team should chime in on this one.

@tommcdon @noahfalk

Reply:
@davmason

Reply:
ping @noahfalk @davmason 
Can you take a look at this part of this change and let me know if this TraceCall check is in the right spot or if it needs to be done earlier?

Reply:
Sorry for the delay, I don't know this code well enough to answer without researching a bit and I haven't had time yet. Noah just got back from vacation and is pretty busy, maybe @hoyosjs knows. Otherwise I'll try to make time soon.

Reply:
I took a look tonight, I think Jan is right here. We should call TraceCall() with the target address just before we invoke the managed code (e.g. as in COMToCLRInvokeTarget and MethodDescCallSite::CallTargetWorker among others).

Calls to these helpers are inserted by the jit in to the prolog/epilog of the managed function, so by the time we are executing JIT_ReversePInvokeEnter we are already in managed code.

Is there an earlier spot you can put the call before it gets to managed code?

Reply:
We are not calling TraceCall on non-x86 today and we have not received any bug reports for it. We should start by creating a repro that demonstrates that there is a problem to fix. If there is one, it may be best to fix it in a separate PR.

Reply:
There is a problem, but it's a pretty narrow case. For the debugger to miss the breakpoint on x64 you have to be in mixed mode debugging and have the reverse pinvoke run on a thread that hasn't had managed code before. I filed #47037 to track it, feel free to punt it here.

Reply:
So we can't actually move this call earlier because this is the earliest point that a managed thread object is set up and we're in COOP GC mode.

Reply:
I've validated that this PR fixes #47037 for the Reverse P/Invoke IL stub case on all platforms, but doesn't fix it for UnmanagedCallersOnly (and as a result, regresses UnmanagedCallersOnly on x86).

Reply:
I think there should be a way to fix this for both cases, without passing the MethodDesc* into the helper.

Reply:
Do you have any ideas off the top of your head?

If not, can we table solving this for UnmanagedCallersOnly for a future PR?

Reply:
I have not stepped through the repro to see what's going on exactly. I am guessing that debugger does not get a chance to setup its stuff on the new thread, and that is why it misses the breakpoint first time through but everything works fine second time through. Is it the case? Can we tell debugger when we create the `Thread*` inside `JIT_ReversePInvokeEnter` that there is a new thread entering via reverse P/Invoke and the debugger should setup its stuff on this new thread?

I agree that it would best for this PR be just about unifying x86 with other platforms, even if it means that x86 will be as broken as other platforms. Then deal with the debugger and profiler issues in separate PRs.

Reply:
I'm not sure which debugger API we should use for that. @davmason any ideas if there's an API we can use here to set up the thread info for the debugger from within JIT_ReversePInvokeEnter when we enter on a new thread?

Reply:
I've validated that we do correctly tell the debugger that we created a new `Thread*` object, so the issue is something else.

Reply:
With the change in 22baef7344f8dd7ebc31321da9b032342e5ecccb, I was able to get the debugger to stop on a breakpoint in an UnmanagedCallersOnly method called from a new native thread. @jkotas let me know if you like the fix.

Reply:
Looks good to me - if it is solving the problem.

Reply:
I’ve validated that it does solve the problem.

Reply:
@jkoritzinsky @jkotas I think I've run into this issue. See my comment below:

https://github.com/dotnet/runtime/issues/47037#issuecomment-777927501

> I believe I am also seeing this exact issue in my new .Net 5 x64 project. I have a thread in my C++ code that sends data back to my .Net UI. I can't debug the reverse Pinvoked .Net method.

It sounds like this issue will be fixed in .Net 6.0. Is that correct?